pkg/speaker/manager.go

```diff
@@ -0,0 +1,415 @@
+	defer m.Unlock()
+	_, exist := m.eips[eip.GetName()]
+	if exist {
+		if !eip.DeletionTimestamp.IsZero() {
+			m.V(3).Info(fmt.Sprintf("deleting eip:[%s]", eip.GetName()))
+			delete(m.eips, eip.GetName())
+		}
+
+		return nil
+	}
+
+	if !eip.DeletionTimestamp.IsZero() {
+		return nil
+	}
+
+	// TODO: layer2 validate NIC infos
+	m.V(3).Info(fmt.Sprintf("store eip:[%s]", eip.GetName()))
+	m.eips[eip.GetName()] = eip
+	return nil
+}
+
+func (m *Manager) HandleService(svc *corev1.Service) error {
+	if svc == nil {
+		return nil
+	}
+
+	svcNSName := types.NamespacedName{Namespace: svc.GetNamespace(), Name: svc.GetName()}.String()
+
+	// get local record info, if localRecord is nil -- no record exists(initing)
+	localRecord, exist := m.ips[svcNSName]
+	if !exist && svc.Spec.Type != corev1.ServiceTypeLoadBalancer {
+		m.V(3).Info(fmt.Sprintf("openelb no record about servic:[%s] type:[%s]", svcNSName, svc.Spec.Type))
+		return nil
+	}
+
+	// deleting svc
+	if !svc.DeletionTimestamp.IsZero() {
+		if localRecord == nil {
+			m.Info(fmt.Sprintf("service:%s is deleting, so don't handler it", svcNSName))
+			return nil
+		}
+		m.V(3).Info(fmt.Sprintf("deleting svc:[%s - %s], so delete LoadBalancer", svc.GetName(), localRecord.String()))
+		return m.delLoadBalancer(svc, localRecord)
+	}
+
+	// update service from TypeLoadBalancer to other type
+	if exist && svc.Spec.Type != corev1.ServiceTypeLoadBalancer {
+		if localRecord == nil {
+			m.Info(fmt.Sprintf("local:%s record exists but no data. changed svc'type is %s", svcNSName, svc.Spec.Type))
+			return nil
+		}
+		m.V(3).Info(fmt.Sprintf("change service:[%s] type from LoadBalancer to %s, so delete LoadBalancer", svc.GetName(), svc.Spec.Type))
+		return m.delLoadBalancer(svc, localRecord)
+	}
+
+	// get service record info
+	svcRecord := m.getSvcRecordInfo(svc)
+	if svcRecord == nil {
+		m.V(3).Info(fmt.Sprintf("get service:[%s] record info error", svc.GetName()))
+		return nil
+	}
+
+	return m.handleSvcBalance(svc, localRecord, svcRecord)
+}
+
+func (m *Manager) handleSvcBalance(svc *corev1.Service, localRecord, svcRecord *recordInfo) error {
+	switch m.getHandleResult(localRecord, svcRecord) {
+	case wantDelete:
+		m.V(1).Info("delLoadBalancer " + localRecord.String())
+		return m.delLoadBalancer(svc, localRecord)
+	case wantStore:
+		m.V(1).Info("setLoadBalancer " + svcRecord.String())
+		return m.setLoadBalancer(svc, svcRecord)
+	case wantReset:
+		m.V(1).Info(fmt.Sprintf("resetLoadBalancer localRecord:%s svcRecord:%s", localRecord.String(), svcRecord.String()))
+		return m.resetLoadBalancer(svc, localRecord, svcRecord)
+
+	default:
+		// endpoint update
+		if svc.Spec.ExternalTrafficPolicy == corev1.ServiceExternalTrafficPolicyTypeLocal {
+			m.V(1).Info("Local svc.externalTrafficPolicy updateLoadBalancer")
+			return m.updateLoadBalancer(svc, svcRecord)
+		}
+		m.V(1).Info("handler do nothing")
+	}
+
+	return nil
+}
+
+func (m *Manager) getHandleResult(localRecord, svcRecord *recordInfo) handleResult {
```

Comment:
/lgtm


We need to add some unit tests, which can be included in the next PR.